libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+    hooks.createOperations.tapPromise(
+      PLUGIN_NAME,
+      async (operations: Set<Operation>, context: ICreateOperationsContext): Promise<Set<Operation>> => {
+        const { buildCacheConfiguration, isIncrementalBuildAllowed, cobuildConfiguration } = context;
+        if (!buildCacheConfiguration) {
+          return operations;
+        }
+
+        let disjointSet: DisjointSet<Operation> | undefined;
+        if (cobuildConfiguration?.cobuildEnabled) {
+          disjointSet = new DisjointSet<Operation>();
+        }
+
+        for (const operation of operations) {
+          disjointSet?.add(operation);
+          const { runner } = operation;
+          if (runner) {
+            const buildCacheContext: IOperationBuildCacheContext = {
+              // ShellOperationRunner supports cache writes by default.
+              isCacheWriteAllowed: true,
+              isCacheReadAllowed: isIncrementalBuildAllowed,
+              isSkipAllowed: isIncrementalBuildAllowed,
+              projectBuildCache: undefined,
+              cobuildLock: undefined,
+              cobuildClusterId: undefined,
+              buildCacheTerminal: undefined,
+              buildCacheProjectLogWritable: undefined
+            };
+            // Upstream runners may mutate the property of build cache context for downstream runners
+            this._buildCacheContextByOperationRunner.set(runner, buildCacheContext);
+
+            if (runner instanceof ShellOperationRunner) {
+              this._applyShellOperationRunner(runner, context);
+            }
+          }
+        }
+
+        if (disjointSet) {
+          // If disjoint set exists, connect build cache disabled project with its consumers
+          await Async.forEachAsync(
+            operations,
+            async (operation) => {
+              const { associatedProject: project, associatedPhase: phase } = operation;
+              if (project && phase && operation.runner instanceof ShellOperationRunner) {
+                const buildCacheEnabled: boolean = await this._tryGetProjectBuildEnabledAsync({
+                  buildCacheConfiguration,
+                  rushProject: project,
+                  commandName: phase.name
+                });
+                if (!buildCacheEnabled) {
+                  for (const consumer of operation.consumers) {
+                    if (consumer.runner instanceof ShellOperationRunner) {
+                      disjointSet?.union(operation, consumer);
+                    }
+                  }
+                }
+              }
+            },
+            {
+              concurrency: 10
+            }
+          );
+
+          for (const set of disjointSet.getAllSets()) {
+            if (cobuildConfiguration?.cobuildEnabled && cobuildConfiguration.cobuildContextId) {
+              // Generates cluster id, cluster id comes from the project folder and phase name of all operations in the same cluster.
```

Comment:
Thanks to the addition of operations to the disjoint set forest happening asynchronously, the order of the operations in any set, and the order of the sets, is not deterministic, which means the cluster ids will not be deterministic. Will that be a problem?

Reply:
You are right. This is really nice catch! The first idea comes to me is to sort operations. But i found it's hardly to do it in a easy way. Let me figure it out.